### PR TITLE
Admin audio recording pages

### LINKED ISSUE #129: Admin Audio Recordings
Create admin audio recordings page (`/admin/audio_recordings`)

### PR DESCRIPTION
**Pull request by Allcharles**
# Admin Audio Recording Pages

Created the list and detail admin audio recording pages.

## Changes

- Extracted generic pagedTableTemplate dependent component tests to `/src/app/test/helpers/pagedTableTemplate.ts` file
- Updated all admin list components to use new generic tests
- Updated `RenderViewComponent` to handle `Observable` and `AbstractModel` inputs

## Issues

- New tests run a little slow, could be an idea in future to optimize them if a lot more list components are built

## Closes

Closes #129 

## Visual Changes

List component:
image

Detail component:
image

### CODE REVIEW

File: src/app/component/admin/audio-recordings/list/list.component.spec.ts [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    for (let i = 0; i < 25; i++) {
+      defaultAudioRecordings.push(
+        new AudioRecording({
+          id: i,
+          siteId: 1,
+          durationSeconds: 3000,
+          recordedDate: "2020-03-09T22:00:50.072+10:00",
+        })
+      );
+    }
```
Comment (atruskie) [inline src/app/component/admin/audio-recordings/list/list.component.spec.ts:45]: If you find you need faked objects a lot, try something like https://github.com/marak/Faker.js/

File: src/app/component/admin/audio-recordings/list/list.component.ts [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    this.sortKeys = {
+      id: "id",
+      site: "siteId",
+      duration: "durationSeconds",
+      recorded: "recordedDate",
+    };
+    this.getPageData();
+  }
+
+  viewRedirectPath(model: AudioRecording) {
```
Comment (atruskie) [inline src/app/component/admin/audio-recordings/list/list.component.ts]: Why is this a `redirect`? What url is gets replaces with this one?

File: src/app/component/admin/scripts/list/list.component.spec.ts [hunk 3]
```diff
@@ -1,23 +1,57 @@
...
+      executableCommand: "command",
+    });
+    defaultModels = [];
+    for (let i = 0; i < 25; i++) {
+      defaultModels.push(
+        new Script({
+          id: i,
+          name: "script " + i,
+          version: 0.1,
+          executableCommand: "command",
```
Comment (atruskie) [inline src/app/component/admin/scripts/list/list.component.spec.ts:42]: second the faker comment

File: src/app/models/AbstractModel.ts [hunk 4]
```diff
@@ -74,6 +74,13 @@ export abstract class AbstractModel {
     return output;
   }
 
+  /**
+   * Convert model to string
+   */
+  public toString(): string {
+    return this.id.toString();
```
Comment (atruskie) [inline src/app/models/AbstractModel.ts]: This could be annoying; how does this look in the debug console?

File: src/app/models/AnalysisJob.ts [hunk 5]
```diff
@@ -105,6 +105,10 @@ export class AnalysisJob extends AbstractModel implements IAnalysisJob {
   public get viewUrl(): string {
     throw new Error("AnalysisJob viewUrl not implemented.");
   }
+
+  public toString(): string {
+    return this.name;
```
Comment (atruskie) [inline src/app/models/AnalysisJob.ts]: If the `id` was important enough to show in the abstract model, it is probably useful here too

**Review by atruskie — APPROVED** [review]
With the `toString` overrides:
- ensure in the debug console the object can still be expanded
- i'd suggest a minimum of always outputing model name type (e.g. `AnalysisJob`) and `id` (e.g. `AnalysisJob:123`)
- can we generalise printing the `name` in Abstract model? TS should be able to do a function that tests if base type has `name`, and returns value if object has field

Only minor nits, good to go

File: src/app/models/AbstractModel.ts (hunk 4 above)
Comment (Allcharles) [reply src/app/models/AbstractModel.ts]: `toString()` does not affect the debug console output. Debug console displays the `toJSON()` method.

File: src/app/component/admin/audio-recordings/list/list.component.spec.ts (hunk 1 above)
Comment (Allcharles) [reply src/app/component/admin/audio-recordings/list/list.component.spec.ts:45]: https://github.com/QutEcoacoustics/workbench-client/issues/77. Its in my backlog, however the need for faker is increasing a bit.

**Review by atruskie — APPROVED** [review]
👌 nice and neat

**Pull request merged**